File: ros2_ws/src/cpp_pubsub_lab3/src/listener.cpp
#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

// 1. 定义一个普通的全局/静态回调函数（不再属于任何类）
void cb(const std_msgs::msg::String::SharedPtr msg) {
    // 这里的日志器需要通过一个特定的宏来获取，或者直接用 std::cout
    RCLCPP_INFO(rclcpp::get_logger("listener_logger"), "[cpp] I heard: %s", msg->data.c_str());
}

int main(int argc, char * argv[]) {
    rclcpp::init(argc, argv);

    // 2. 创建纯节点对象
    auto node = rclcpp::Node::make_shared("listener_cpp_lab3");

    // 3. 创建订阅者，直接传入普通函数指针 'cb'
    // 不需要写令人头疼的 std::bind 和 std::placeholders
    auto sub = node->create_subscription<std_msgs::msg::String>("chatter", 10, cb);

    // 4. 进入阻塞循环，等待接收消息
    rclcpp::spin(node);

    rclcpp::shutdown();
    return 0;
}

File: submit/ros2_ws/src/py_pubsub_lab3/py_pubsub_lab3/listener.py
#!/usr/bin/env python
import rclpy
from rclpy.node import Node
from std_msgs.msg import String

class ListenerPyLab3(Node):
    def __init__(self):
        super().__init__('listener_py_lab3')

        self.subscription = self.create_subscription(
            String,
            "chatter",
            self.cb,
            10
        )

    def cb(self,msg):
        self.get_logger().info(f'f[py] I  heard:{msg.data}')

def main(args = None):
    rclpy.init(args=args)
    node = ListenerPyLab3()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()

if __name__ == "__main__":
    main()